lib/rucio/core/replica.py

```diff
@@ -2733,3 +2733,26 @@ def get_suspicious_files(rse_expression, younger_than=None, nattempts=None, sess
             rses[rse_id] = rse
         result.append({'scope': scope, 'name': name, 'rse': rses[rse_id], 'cnt': cnt, 'created_at': created_at})
     return result
+
+
+@transactional_session
+def set_tombstone(rse, scope, name, rse_id=None, session=None):
+    """
+    Sets a tombstone on a replica.
+
+    :param rse: name of the RSE.
+    :param scope: scope of the replica DID.
+    :param name: name of the replica DID.
+    :param rse_id: optional ID of RSE.
+    :param session: database session in use.
+    """
+    if not rse_id:
+        rse_id = get_rse_id(rse)
+    replica = session.query(models.RSEFileAssociation).filter_by(rse_id=rse_id, name=name, scope=scope).first()
+    if not replica:
+        raise exception.ReplicaNotFound('Replica %s:%s on RSE %s could not be found.' % (scope, name, rse))
+    if not replica.lock_cnt:
+        replica.tombstone = OBSOLETE
+        replica.save(session=session)
```

Comment:
I think we should make it sure and really check the locks table here as well, just to make sure that there are no locks and that the lock_cnt is not inconsistent. Could also be done in one query with a join.

Reply:
okay, shall I also change to one query? Makes sense to me as its saving interactions. But then I cannot distinguish if  the replica was not found or if its locked in case no rows get updated. Maybe useful information for error handling on the client side.

Reply:
`update table where not exists (select 1 from another_table) etc` might work. 
If the update returns 0 rows affected, one can check the replicas presence, etc to give back a proper error message. It's still performant.